tools/json_schemas_to_typescript.rb

```diff
@@ -21,64 +21,116 @@ def initialize_listener(app)
-
-            export type #{type_name} = FromSchema<typeof schema>;
-          JS
-
-          File.write(file_to_write, typescript_content)
+          if write_all || schema_path.in?(changed) || schema_path.in?(added)
+            response_name =
+              schema_path.
+                gsub(SCHEMA_DIRECTORY_REGEX, '').
+                delete_suffix('.json').
+                split('/').
+                then do |path_fragments|
+                  if path_fragments[-1] == 'index'
+                    [*path_fragments[0..-3].map(&:singularize), *path_fragments[-2..]]
+                  else
+                    path_fragments.map(&:singularize)
+                  end
+                end.
+                join('/').
+                camelize.
+                gsub('::', '')
+
+            type_name = "#{response_name}#{type_suffix}"
+
+            file_to_write = "#{types_directory}/#{type_name}.ts"
+
+            system(<<~SH.squish)
+              ./node_modules/.bin/quicktype
+                --src-lang schema
+                #{schema_path}
+                --out #{file_to_write}
+                --just-types
+            SH
+
+            # Only export the first interface in the file. (All are exported by quicktype.)
+            # Swap `Date` to `string`.
+            content_to_write =
+              File.read(file_to_write).
+                gsub(/^export interface/, 'interface').
+                sub(/^interface/, 'export interface').
+                gsub(/\bDate\b/, 'string')
+
+            content_to_write =
+              <<~CONTENT
+                // This is a generated file. Do not edit this file directly.
+
+                #{content_to_write.rstrip}
+              CONTENT
+
+            File.write(
+              file_to_write,
+              content_to_write,
+            )
+
+            # Work around this quicktype bug:
+            # https://github.com/glideapps/quicktype/issues/ 2481
+            if JSON.parse(File.read(schema_path))['type'] == 'array'
+              file_content = File.read(file_to_write)
+
+              main_interface =
+                file_content.
+                  match(/^export interface (?<main_interface>\w+) {/).
+                  []('main_interface')
+
+              main_interface_element_name = "#{main_interface}Element"
+
+              file_content.gsub!(
+                /^export interface #{main_interface} {/,
+                "interface #{main_interface_element_name} {",
+              )
+
+              File.write(file_to_write, <<~JS)
+                export type #{main_interface} = Array<#{main_interface_element_name}>
+
+                #{file_content.rstrip}
+              JS
+            end
+          end
         end
       end
     end
-    # rubocop:enable Metrics/MethodLength
+    # rubocop:enable Metrics
+
+    private
+
+    def relative_path(absolute_path)
+      Pathname.new(absolute_path).relative_path_from(Rails.root).to_s
+    end
```

Comment:
The DRY zealot in me wants to extract this to some shared location. We do basically the same thing here: https://github.com/davidrunger/david_runger/blob/7b24cbeb0156486c4ae84d3cace58bf1bd7bbbcc/lib/generators/datamigration/datamigration_generator.rb/#L17 .